xls/modules/fp_trig_reduce_64_test.cc

```diff
@@ -0,0 +1,261 @@
+    0xb6c52b3278872083,
+    0xfca2c757bd778ac3,
+    0x6e48dc74849ba5c0,
+    0x0c925dd413a32439,
+    0xfc3bd63962534e7d,
+    0xd1046bea5d768909,
+    0xd338e04d68befc82,
+    0x7323ac7306a673e9,
+    0x3908bf177bf25076,
+    0x3ff12fffbc0b301f,
+    0xde5e2316b414da3e,
+    0xda6cfd9e4f96136e,
+    0x9e8c7ecd3cbfd45a,
+    0xea4f758fd7cbe2f6,
+    0x7a0e73ef14a525d4,
+    0xd7f6bf623f1aba10,
+    0xac06608df8f6d757,
+  };
+
+  constexpr uint64_t FP64_EXP_SHIFT = 52;
+  constexpr uint64_t FP64_EXP_BIAS = 1023;
+  constexpr uint64_t FP64_EXP_MASK = 0x7ff;
+
+
+  if (x < M_PI_4) {
+    return ResultT{.fraction = x, .j = static_cast<char>(0)};
+  }
+
+  // Extract out the integer and exponent such that,
+	// x = ix * 2 ** exp.
+  uint64_t ix = flatten(x);
+  int exp = int((ix>>FP64_EXP_SHIFT)&FP64_EXP_MASK) - FP64_EXP_BIAS - FP64_EXP_SHIFT;
+  ix  = ~(FP64_EXP_MASK << FP64_EXP_SHIFT) & ix;
+  ix |= uint64_t(1) << FP64_EXP_SHIFT;
+
+  // Use the exponent to extract the 3 appropriate uint64 digits from MPI4,
+  uint64_t digit = uint64_t(exp+61)/64;
+  uint64_t bitshift = uint64_t(exp+61)%64;
+
+  uint64_t z0 = (MPI4[digit] << bitshift) | 
+                  // It turns out that a shift of size >= type size (e.g. 64)
+                  // is undefined behavior in C. Bafflingly, uint64_t << 64
+                  // result in 0 if the shift amount is statically determined
+                  // but results in no shift if the shift ammount is dynamically
+                  // determined...
+                  (bitshift == 0 ? 0 : (MPI4[digit + 1] >> (64 - bitshift)));
+  uint64_t z1 = (MPI4[digit + 1] << bitshift) |
+                  (bitshift == 0 ? 0 : (MPI4[digit + 2] >> (64 - bitshift)));
+  uint64_t z2 = (MPI4[digit + 2] << bitshift) |
+                  (bitshift == 0 ? 0 : (MPI4[digit + 3] >> (64 - bitshift)));
+
+  // Multiply mantissa by the digits and extract the upper two digits (hi, lo).
+  absl::uint128 z2_prod = static_cast<absl::uint128>(z2) * static_cast<absl::uint128>(ix);
+  uint64_t z2hi = absl::Uint128High64(z2_prod);
+  absl::uint128 z1_prod = static_cast<absl::uint128>(z1) * static_cast<absl::uint128>(ix);
+  uint64_t z1hi = absl::Uint128High64(z1_prod);
+  uint64_t z1lo = absl::Uint128Low64(z1_prod);
+  uint64_t z0lo = z0 * ix;
+
+  // We probably don't have to break up z1 to do this, but since this 
+  // code is just for testing, we will follow the source as closely as possible
+  absl::uint128 z_sum = static_cast<absl::uint128>(z1lo) 
+                            + static_cast<absl::uint128>(z2hi)
+                            + (static_cast<absl::uint128>(z1hi) << 64)
+                            + (static_cast<absl::uint128>(z0lo) << 64);
+  uint64_t hi = absl::Uint128High64(z_sum);
+  uint64_t lo = absl::Uint128Low64(z_sum);
+
+  // The top 3 bits are j.
+  uint8_t j = hi >> 61;
+  // Extract the fraction and find its magnitude.
+  hi = hi << 3 | lo >> 61;
+  // Note: This shift is not present in the go implementation, but
```

Comment:
Do you have a repro for the incorrect case in the Go library? Might be worth filing a bug if you're feeling ambitious.

Reply:
repro?